packages/sourcecred/src/core/ledger/ledger.js

```diff
@@ -847,6 +869,153 @@ export class Ledger {
+  }
+
+  /**
+   * Disabling accounting has the following effects on the ledger:
+   * - disables grain transfers
+   * - zeroes out all balances
+   * - deactivates accounts that don't have a PayoutAddress set for the
+   *   current instance configuration, and prevents them from activating until
+   *   they do.
+   *
+   *   This function can be called while accounting is disabled to update
+   *   the ledger's external currency configuration. Note that this will also
+   *   deactivate identities missing a payout address that matches the new
+   *   configuration.
+   */
+  disableAccounting(): Ledger {
+    // ensure an external currency is set
+    this.externalCurrency();
+    // ensure accounting isn't already disabled
+    if (this._balanceAccountingEnabled) {
+      this._createAndProcessEvent({
+        type: "DISABLE_ACCOUNTING",
+      });
+    }
+    // no-op if already disabled
+    return this;
+  }
+
+  _disableAccounting(_: DisableAccounting) {
+    this._balanceAccountingEnabled = false;
+
+    const currencyKey = getCurrencyKey(this.externalCurrency());
+
+    // Warning! Mutations below
+    const accountsIterator = this._accounts.values();
+    for (const account of accountsIterator) {
+      // zero out all balances, so there is no confusion about them.
+      // By zeroing them out, they cannot be compared to external balances
+      // in any meaningful way.
+      account.balance = G.ZERO;
+      // deactivate accounts that don't have a payout address assigned
+      // recursive operations are occuring here. Need to be careful with
+      // these next op calls
+      const payoutAddress = account.payoutAddresses.get(currencyKey);
+      if (!payoutAddress) this.deactivate(account.identity.id);
+    }
+  }
+
+  /**
+   * Disabling accounting has the following effects on the ledger:
+   * - disables grain transfers
+   * - zeroes out all balances
+   * - deactivates accounts that don't have a PayoutAddress set for the
+   *   current instance configuration, and prevents them from activating until
+   *   they do.
+   *
+   *   This function can be called while accounting is disabled to update
+   *   the ledger's external currency configuration. Note that this will also
+   *   deactivate identities missing a payout address that matches the new
+   *   configuration.
+   */
+  setExternalCurrency(chainId: ChainId, tokenAddress?: EthAddress): Ledger {
+    // ensure accounting isn't already disabled
+    const currency = buildCurrency(chainId, tokenAddress);
+
+    if (!deepEqual(currency, this._externalCurrency)) {
+      this._createAndProcessEvent({
+        type: "SET_EXTERNAL_CURRENCY",
+        currency,
+      });
+
+      if (!this._balanceAccountingEnabled) {
+        // Deactivate accounts if they don't have an address stored for
+        // the newly configured currency.
+        this._createAndProcessEvent({
+          type: "DISABLE_ACCOUNTING",
+        });
+      }
```

Comment:
This should go in _setExternalCurrency

Non-underscored: utility wrapper for a single action (absolutely no mutation beyond a single _createAndProcessEvent, virtually no business logic, maybe small efficiency no-op logic)
Underscored: strongly coded and semantically whole action (self-sufficient business logic and validation)